src/Razor/src/Microsoft.VisualStudio.DevKit.Razor/DevKitTelemetryReporter.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.AspNetCore.Razor.Telemetry;
+using Microsoft.VisualStudio.Telemetry;
+
+#if DEBUG
+using System.Linq;
+#endif
+
+namespace Microsoft.VisualStudio.DevKit.Razor;
+
+[Shared]
+[Export(typeof(IDevKitTelemetryReporter))]
+internal sealed class DevKitTelemetryReporter : IDevKitTelemetryReporter
+{
+    private TelemetrySession? _telemetrySession;
+    private const string CollectorApiKey = "0c6ae279ed8443289764825290e4f9e2-1a736e7c-1324-4338-be46-fc2a58ae4d14-7255";
+
+    public void InitializeSession(string telemetryLevel, string? sessionId, bool isDefaultSession)
+    {
+        Debug.Assert(_telemetrySession == null);
+
+        var sessionSettingsJson = CreateSessionSettingsJson(telemetryLevel, sessionId);
+        var session = new TelemetrySession($"{{{sessionSettingsJson}}}");
+
+        if (isDefaultSession)
+        {
+            TelemetryService.SetDefaultSession(session);
+        }
+
+        session.Start();
+        session.RegisterForReliabilityEvent();
+
+        _telemetrySession = session;
+    }
+
+    public IDisposable BeginBlock(string name, Severity severity)
+    {
+        return BeginBlock(name, severity, ImmutableDictionary<string, object?>.Empty);
+    }
+
+    public IDisposable BeginBlock(string name, Severity severity, ImmutableDictionary<string, object?> values)
+    {
+        return new TelemetryScope(this, name, severity, values.ToImmutableDictionary((tuple) => tuple.Key, (tuple) => tuple.Value));
+    }
+
+    public void ReportEvent(string name, Severity severity)
+    {
+        var telemetryEvent = new TelemetryEvent(TelemetryHelpers.GetTelemetryName(name), TelemetryHelpers.ToTelemetrySeverity(severity));
+        Report(telemetryEvent);
+    }
+
+    public void ReportEvent(string name, Severity severity, ImmutableDictionary<string, object?> values)
+    {
+        var telemetryEvent = new TelemetryEvent(TelemetryHelpers.GetTelemetryName(name), TelemetryHelpers.ToTelemetrySeverity(severity));
```

Comment:
Can the body of this method be moved to TelemetryHelpers too?

Also Report, ReportFault, TrackLspRequest etc. They feel very similar, but I didn't do a line-by-line check :)

Reply:
Yea, if we already have shared code we might as well just do inheritance to cover this.